### PR TITLE
484 compliance info

### LINKED ISSUE #484: Compliance
## Formalizing the FDC3 Standard documentation

This issue has been raised as part of work, sponsored by FINOS, to improve and formalize the documentation of the FDC3 standard.

##  Issue description

“FDC3 Compliance” states, “… So, **unless a particular item in a spec is marked with keywords such as OPTIONAL, MAY, SHOULD, or SHOULD NOT, it should be treated as REQUIRED.** Since FDC3 itself is primarily concerned with establishing the baseline requirements for interoperation, this is consistent with the IETF Guidance:”

A search of all docs/md files shows that “MUST” appears only three times, and “SHALL” only once, which reinforces that the advice written above has been followed.

**My first, second, and third reactions to this approach were negative.** These keywords exist to mark specific text according to the keyword’s property, yet there are almost *never* used for that purpose. I would have expected to see *all required criteria* to be *explicitly* marked “SHALL” rather than that being the default. Without that, if a paragraph or section does *not* contain any of these keywords, the reader still has to scan it *very* carefully to see what requirements might be contained therein.

**Proposals:**

1. Rewrite this section, saying that we use the RFC-2119 required keywords SHALL/MUST instead of defaulting to them. That is, we make all requirements explicit. (This allows them to be found mechanically to help the implementer figure out exactly what they have to do to comply.)
1. Apply the above-mentioned change throughout the Parts.
1. Define the term *compliant implementation* in terms of which Parts are required and which, if any, are optional.
1. Introduce the notion of *deprecation* (which is used at least in “API Specification, Raising Intents”).

### LINKED ISSUE #636: Provide details about support for optional API features in `ImplementationMetadata`
### Use Case:
One of the goals of the FDC3 standard is to allow apps from vendors to interoperate with each other in a standardized fashion, via a compliant implementation of the Desktop Agent API. However, as the standard can contain optional features (indicated via the SHOULD or MAY keywords) it would be helpful to define, as part of the standard, how support for these features can be queried or tested.

### Proposal
The `fdc3.getInfo()` call added in 1.2 returns an [`ImplementationMetadata` object](https://fdc3.finos.org/docs/api/ref/Metadata#implementationmetadata) that contains information about the FDC3 standard version and Desktop Agent implementation and would seem an ideal place to also insert details of which optional features are supported, e.g..

```typescript
enum OptionalFeatures {
   "USER_CHANNEL_PICKER": "UserChannelPicker",
   ... // further optional features
}

public interface ImplementationMetadata {
  fdc3Version: string;
  provider: string;
  providerVersion?: string;
  optionalFeatureSupport: Record<OptionalFeatures, boolean>;
}
```

#### Usage example: 
```javascript
//Check for an optional feature
if (fdc3.getInfo) {
  let implementationMetadata = await fdc3.getInfo();
  if (!implementationMetadata.optionalFeatureSupport[OptionFeatures.USER_CHANNEL_PICKER]) {
      //render our own channel picker as the Desktop Agent doesn't provide one
  }
} else {
  //fdc3 version < 1.2 or non-compliant implementation
}
```

### Additional information
Optional features can and should be added to the enumeration alongside the work on #484 to formalize compliance information throughout the standard

### PR DESCRIPTION
**Pull request by kriswest**
resolves #484 
resolves #636 

Collated compliance information for the 4 parts of the standard. Does not change the standard, only the documentation to collate existing requirements. Approved to merge post-adoption of 2.0 by SWG at #728. Hence, it only requires maintainer and editor review to merge.

Preview links to check out:
- [Compliance page personas](https://deploy-preview-753--lambent-kulfi-cf51a7.netlify.app/docs/next/fdc3-compliance#personas) - these used to have a subset of compliance reqs underneath them - these now link out to more detailed compliance details in the individual specs
- Compliance basically breaks down into Platform providers needing to comply with the two API specs, while App providers need to comply with the intents and contexts specs:
  - [Desktop Agent API compliance](https://deploy-preview-753--lambent-kulfi-cf51a7.netlify.app/docs/next/api/spec#desktop-agent-api-standard-compliance)
  - [App Directory API compliance](https://deploy-preview-753--lambent-kulfi-cf51a7.netlify.app/docs/next/app-directory/spec#app-directory-standard-compliance)
  - [Intents compliance](https://deploy-preview-753--lambent-kulfi-cf51a7.netlify.app/docs/next/intents/spec#intents-standard-compliance)
  - [Context compliance](https://deploy-preview-753--lambent-kulfi-cf51a7.netlify.app/docs/next/context/spec#context-data-standard-compliance)

### CODE REVIEW

**Review by RexJaeschke — APPROVED** [review]

**Review by rikoe — APPROVED** [review]

File: docs/api/spec.md [hunk 1]
```diff
@@ -105,6 +92,63 @@ Desktop Agent interop is supported by common standards for APIs for App discover
 
 An actual connection protocol between Desktop Agents is not currently available in the FDC3 standard, but work on creating one for a future version of the standard is underway (see [GitHub discussion #544](https://github.com/finos/FDC3/discussions/544) for details).
 
+### Desktop Agent API Standard Compliance
+
+An FDC3 Standard compliant Desktop Agent implementation **MUST**:
+
+- Provide the FDC3 API to web applications via a global accessible as [`window.fdc3`](#api-access).
+- Provide a global [`fdc3Ready`](#api-access) event that is fired when the API is ready for use.
+- Provide a method of resolving ambiguous intents (i.e. those that might be resolved by multiple applications) or unspecified intents (calls to raiseIntentForCOntext that return multiple options), usch as a resolver UI.
```
Comment (greyseer256) [inline docs/api/spec.md:101]:
```suggestion
- Provide a method of resolving ambiguous intents (i.e. those that might be resolved by multiple applications) or unspecified intents (calls to raiseIntentForContext that return multiple options), such as a resolver UI.
```

File: docs/context/spec.md [hunk 2]
```diff
@@ -135,31 +135,50 @@ E.g. `"CURRENCY_ISOCODE": "GBP"`
...
+## Context Data Standard Compliance
+
+An FDC3 Standard compliant application that supports the use of context data **MUST**:
+
+* Ensure that any FDC3-defined standard context types used meet the interface defined for that type of context data.
+* Where Channels are supported or an app is intended to receive context from [`fdc3.open`](api/ref/DesktopAgent#open) calls, use the [`fdc3.addContextListener`](api/ref/DesktopAgent#addcontextlistener) API call to set up appropriate handlers on start-up (for User channels and for receiving context from [`fdc3.open`](api/ref/DesktopAgent#open)) or when the channel is first created or retrieved (for App and Private channels).
+
+An FDC3 Standard compliant application that supports the use of context data **SHOULD**:
+
+* Prefer FDC3-defined standard context types over proprietary contexts, where a suitable FDC3-defined stanrd context type is available.
```
Comment (greyseer256) [inline docs/context/spec.md:147]:
```suggestion
* Prefer FDC3-defined standard context types over proprietary contexts, where a suitable FDC3-defined standard context type is available.
```

File: docs/context/spec.md [hunk 3]
```diff
@@ -135,31 +135,50 @@ E.g. `"CURRENCY_ISOCODE": "GBP"`
...
+* Ensure that any proprietary context data types fefined follow any the recommended [namespacing](#namespacing) and [field type conventions](#field-type-conventions) in the specification.
```
Comment (greyseer256) [inline docs/context/spec.md:148]:
```suggestion
* Ensure that any proprietary context data types defined follow any the recommended [namespacing](#namespacing) and [field type conventions](#field-type-conventions) in the specification.
```

File: docs/fdc3-compliance.md [hunk 4]
```diff
@@ -9,48 +9,44 @@ In general, the ratified FDC3 specs represent a lowest common denominator interf
...
+>    example, they must not be used to try to impose a particular method
+>    on implementors where the method is not required for 
+>    interoperability.
 
 These rules would apply only to standards work within FDC3. Today, this covers the API, App Directory, Context Data, and Intents specifications.
 
 ## Personas
-FDC3 implementors generally fall into 2 categories: platform providers, and application providers. A platform provider supplies an implementation of the FDC3 API for applications to use. Implicitly, it connects to one or more App Directories.
+
+FDC3 implementors generally fall into 2 categories: platform providers, and application providers. A platform provider supplies an implementation(s) of the FDC3 APIs (The Desktop Agent API and Applkication Directory) for applications to use.
```
Comment (greyseer256) [inline docs/fdc3-compliance.md:24]:
```suggestion
FDC3 implementors generally fall into 2 categories: platform providers, and application providers. A platform provider supplies an implementation(s) of the FDC3 APIs (The Desktop Agent API and Application Directory) for applications to use.
```

**Review by greyseer256 — APPROVED** [review]
A few typos, content looks good.

File: docs/app-directory/spec.md [hunk 5]
```diff
@@ -16,4 +16,30 @@ View the [full specification](/schemas/next/app-directory) in [OpenAPI v3.0](htt
...
+## App Directory Standard Compliance
+
+An FDC3 Standard compliant App Directory implementation **MUST**:
+
+* Implement the specified `/v2` endpoints for retrieving app definitions as defined in the [app directory OpenAPI specification](/schemas/next/app-directory#tag/Application/paths/~1v2~1apps~1{appId}/get):
+  * `/v2/apps` (GET)
+  * `/v2/apps/{appId}` (GET)
+* Ensure that `appId` field values assigned to applications are unique within the directory.
+* Ensure that app directory records served meet the minimum requirements specified in the [app directory OpenAPI specification](/schemas/next/app-directory#tag/Application/paths/~1v2~1apps~1{appId}/get)
+* Support retrieval of app directory records via either the raw appId (e.g. `myAppId`) or fully-qualified appId (e.g. `myAppId@host.domain.com`) as defined in the [app directory overview](overview#shrinking-the-uri).
```
Comment (kriswest) [inline docs/app-directory/spec.md:30]:
```suggestion
* Support retrieval of app directory records via either the raw `appId` (e.g. `myAppId`) or fully-qualified appId (e.g. `myAppId@host.domain.com`) as defined in the [app directory overview](overview#shrinking-the-uri).
```

**Comment by kriswest** [comment]
> A few typos, content looks good.

Your eyes are sharper than my chubby fingers clearly - many thanks @greyseer256

**Pull request merged**